lib/Conversion/SecretToBGV/SecretToBGV.cpp

```diff
@@ -147,12 +150,80 @@ class SecretGenericOpMulConversion
+  LogicalResult matchAndRewriteInner(
+      secret::GenericOp op, TypeRange outputTypes, ValueRange inputs,
+      ConversionPatternRewriter &rewriter) const override {
+    auto plaintextValues =
+        llvm::to_vector(llvm::make_filter_range(inputs, [&](Value input) {
+          return !isa<lwe::RLWECiphertextType>(input.getType());
+        }));
+    if (!plaintextValues.empty()) {
+      return failure();
+    }
     rewriter.replaceOpWithNewOp<bgv::RelinearizeOp>(
         op, rewriter.create<bgv::MulOp>(op.getLoc(), inputs),
         rewriter.getDenseI32ArrayAttr({0, 1, 2}),
         rewriter.getDenseI32ArrayAttr({0, 1}));
+    return success();
+  }
+};
+
+template <typename T, typename Y>
+class SecretGenericOpCipherConversion : public SecretGenericOpConversion<T, Y> {
+ public:
+  using SecretGenericOpConversion<T, Y>::SecretGenericOpConversion;
+
+  LogicalResult matchAndRewriteInner(
+      secret::GenericOp op, TypeRange outputTypes, ValueRange inputs,
+      ConversionPatternRewriter &rewriter) const override {
+    auto plaintextValues =
+        llvm::to_vector(llvm::make_filter_range(inputs, [&](Value input) {
+          return !isa<lwe::RLWECiphertextType>(input.getType());
+        }));
+    if (!plaintextValues.empty()) {
+      return failure();
+    }
+    rewriter.replaceOpWithNewOp<Y>(op, outputTypes, inputs);
+    return success();
+  }
+};
+
+template <typename T, typename Y>
+class SecretGenericOpCipherPlainConversion
+    : public SecretGenericOpConversion<T, Y> {
+ public:
+  using SecretGenericOpConversion<T, Y>::SecretGenericOpConversion;
+
+  LogicalResult matchAndRewriteInner(
+      secret::GenericOp op, TypeRange outputTypes, ValueRange inputs,
+      ConversionPatternRewriter &rewriter) const override {
+    auto plaintextValues =
+        llvm::to_vector(llvm::make_filter_range(inputs, [&](Value input) {
+          return !isa<lwe::RLWECiphertextType>(input.getType());
+        }));
+    if (plaintextValues.size() != 1) {
+      return failure();
+    }
+
+    TypedValue<RankedTensorType> cleartext;
+    TypedValue<lwe::RLWECiphertextType> ciphertext;
+
+    if (inputs[0] == plaintextValues[0]) {
+      cleartext = cast<TypedValue<RankedTensorType>>(inputs[0]);
+      ciphertext = cast<TypedValue<lwe::RLWECiphertextType>>(inputs[1]);
+    } else {
+      cleartext = cast<TypedValue<RankedTensorType>>(inputs[1]);
+      ciphertext = cast<TypedValue<lwe::RLWECiphertextType>>(inputs[0]);
+    }
+
+    auto plaintextTy = lwe::RLWEPlaintextType::get(
+        op.getContext(), ciphertext.getType().getEncoding(),
+        ciphertext.getType().getRlweParams().getRing(), cleartext.getType());
+    auto plaintext = rewriter.create<lwe::RLWEEncodeOp>(
+        op.getLoc(), plaintextTy, cleartext, ciphertext.getType().getEncoding(),
+        ciphertext.getType().getRlweParams().getRing());
+    rewriter.replaceOpWithNewOp<Y>(op, ciphertext, plaintext);
```

Comment:
It just occurred to me that we don't specify the semantics of SubPlain: is it the ciphertext minus the plaintext or vice versa? I will file an issue.

Reply:
Following up on the issue!